File: src-tauri/src/models/export.rs
use serde::{Deserialize, Serialize};

/// Configuration for video export operation
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportConfig {
    /// Output file path selected by user
    pub output_path: String,

    /// Optional resolution override (width, height)
    /// If None, uses source resolution
    pub resolution: Option<(u32, u32)>,

    /// Optional video codec override
    /// If None, defaults to H.264 (libx264)
    pub codec: Option<String>,

    /// Optional audio codec override
    /// If None, defaults to AAC
    pub audio_codec: Option<String>,

    /// Optional video bitrate (e.g., "5M" for 5 Mbps)
    /// If None, uses FFmpeg default
    pub video_bitrate: Option<String>,

    /// Optional audio bitrate (e.g., "192k")
    /// If None, uses FFmpeg default (128k)
    pub audio_bitrate: Option<String>,
}

impl Default for ExportConfig {
    fn default() -> Self {
        Self {
            output_path: String::new(),
            resolution: None,
            codec: Some("libx264".to_string()),
            audio_codec: Some("aac".to_string()),
            video_bitrate: None,
            audio_bitrate: Some("192k".to_string()),
        }
    }
}

/// Export progress status
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum ExportStatus {
    /// Export is currently running
    Running,

    /// Export completed successfully
    Completed,

    /// Export failed with error
    Failed,

    /// Export was cancelled by user
    Cancelled,
}

/// Progress information for ongoing export
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportProgress {
    /// Export ID
    pub export_id: String,

    /// Current status
    pub status: ExportStatus,

    /// Percentage complete (0-100)
    pub percentage: f32,

    /// Estimated time remaining in seconds
    /// None if not yet calculable
    pub eta_seconds: Option<u64>,

    /// Current processing time in milliseconds
    pub current_time_ms: Option<u64>,

    /// Total timeline duration in milliseconds
    pub total_duration_ms: u64,

    /// Optional error message if status is Failed
    pub error_message: Option<String>,
}

impl ExportProgress {
    /// Create a new export progress tracker
    pub fn new(export_id: String, total_duration_ms: u64) -> Self {
        Self {
            export_id,
            status: ExportStatus::Running,
            percentage: 0.0,
            eta_seconds: None,
            current_time_ms: Some(0),
            total_duration_ms,
            error_message: None,
        }
    }

    /// Update progress from FFmpeg time output
    pub fn update_from_time(&mut self, time_ms: u64, elapsed_seconds: u64) {
        self.current_time_ms = Some(time_ms);

        // Calculate percentage
        if self.total_duration_ms > 0 {
            self.percentage = ((time_ms as f64 / self.total_duration_ms as f64) * 100.0) as f32;
            self.percentage = self.percentage.min(100.0); // Cap at 100%
        }

        // Calculate ETA based on encoding speed
        if time_ms > 0 && elapsed_seconds > 0 {
            let encoding_speed = time_ms as f64 / (elapsed_seconds as f64 * 1000.0);
            let remaining_ms = self.total_duration_ms.saturating_sub(time_ms);
            self.eta_seconds = Some((remaining_ms as f64 / (encoding_speed * 1000.0)) as u64);
        }
    }

    /// Mark export as completed
    pub fn mark_completed(mut self) -> Self {
        self.status = ExportStatus::Completed;
        self.percentage = 100.0;
        self.eta_seconds = Some(0);
        self
    }

    /// Mark export as failed
    pub fn mark_failed(mut self, error: String) -> Self {
        self.status = ExportStatus::Failed;
        self.error_message = Some(error);
        self
    }

    /// Mark export as cancelled
    pub fn mark_cancelled(mut self) -> Self {
        self.status = ExportStatus::Cancelled;
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_export_config_default() {
        let config = ExportConfig::default();
        assert_eq!(config.codec, Some("libx264".to_string()));
        assert_eq!(config.audio_codec, Some("aac".to_string()));
        assert_eq!(config.audio_bitrate, Some("192k".to_string()));
    }

    #[test]
    fn test_export_progress_percentage_calculation() {
        let mut progress = ExportProgress::new("test-id".to_string(), 10000);

        progress.update_from_time(5000, 10);
        assert_eq!(progress.percentage, 50.0);

        progress.update_from_time(10000, 20);
        assert_eq!(progress.percentage, 100.0);

        // Should cap at 100%
        progress.update_from_time(15000, 30);
        assert_eq!(progress.percentage, 100.0);
    }

    #[test]
    fn test_export_progress_status_transitions() {
        let progress = ExportProgress::new("test-id".to_string(), 10000);
        assert_eq!(progress.status, ExportStatus::Running);

        let progress = progress.mark_completed();
        assert_eq!(progress.status, ExportStatus::Completed);
        assert_eq!(progress.percentage, 100.0);

        let progress = ExportProgress::new("test-id".to_string(), 10000);
        let progress = progress.mark_failed("Test error".to_string());
        assert_eq!(progress.status, ExportStatus::Failed);
        assert_eq!(progress.error_message, Some("Test error".to_string()));

        let progress = ExportProgress::new("test-id".to_string(), 10000);
        let progress = progress.mark_cancelled();
        assert_eq!(progress.status, ExportStatus::Cancelled);
    }
}

File: src/lib/tauri/export.ts
import { invoke } from "@tauri-apps/api/core";
import { save } from "@tauri-apps/plugin-dialog";
import type { Timeline } from "@/types/timeline";

/**
 * Export configuration options
 */
export interface ExportConfig {
  outputPath: string;
  resolution?: [number, number]; // [width, height]
  codec?: string; // Default: "libx264"
  audioCodec?: string; // Default: "aac"
  videoBitrate?: string; // e.g., "5M"
  audioBitrate?: string; // e.g., "192k"
}

/**
 * Export progress status
 */
export type ExportStatus = "running" | "completed" | "failed" | "cancelled";

/**
 * Export progress information
 */
export interface ExportProgress {
  exportId: string;
  status: ExportStatus;
  percentage: number; // 0-100
  etaSeconds?: number; // Estimated time remaining
  currentTimeMs?: number; // Current processing time
  totalDurationMs: number; // Total timeline duration
  errorMessage?: string; // Error message if status is "failed"
}

/**
 * Start a new video export operation
 *
 * This function opens a native macOS save dialog, then initiates
 * the export of a timeline to an MP4 file. The export runs in the background.
 *
 * @param timeline - Timeline with tracks and clips to export
 * @param config - Optional export configuration (codec, resolution, etc.)
 * @returns Promise<string> - Export ID for tracking progress, or null if cancelled
 * @throws Error with user-friendly message if export fails to start
 */
export async function startExport(
  timeline: Timeline,
  config?: Partial<ExportConfig>
): Promise<string | null> {
  try {
    // Open native save dialog
    const outputPath = await save({
      title: "Export Video",
      defaultPath: "export.mp4",
      filters: [
        {
          name: "MP4 Video",
          extensions: ["mp4"],
        },
      ],
    });

    // User cancelled the dialog
    if (!outputPath) {
      return null;
    }

    // Build export config with defaults
    const exportConfig: ExportConfig = {
      outputPath,
      codec: config?.codec,
      audioCodec: config?.audioCodec,
      resolution: config?.resolution,
      videoBitrate: config?.videoBitrate,
      audioBitrate: config?.audioBitrate,
    };

    // Debug: Log timeline structure before export
    console.log("[Export Debug] Timeline structure:", {
      trackCount: timeline.tracks.length,
      totalDuration: timeline.totalDuration,
      tracks: timeline.tracks.map((track, idx) => ({
        index: idx,
        id: track.id,
        trackType: track.trackType,
        clipCount: track.clips.length,
        clips: track.clips.map(clip => ({
          id: clip.id,
          filePath: clip.filePath,
          startTime: clip.startTime,
          duration: clip.duration,
        })),
      })),
    });

    // Start export
    const exportId = await invoke<string>("cmd_start_export", {
      timeline,
      config: exportConfig,
    });

    return exportId;
  } catch (error) {
    // Rust backend returns user-friendly error messages as strings
    if (typeof error === "string") {
      throw new Error(error);
    }
    // Fallback for unexpected error types
    throw new Error(
      `Failed to start export: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

/**
 * Get progress information for an ongoing export
 *
 * @param exportId - ID of the export to query (returned from startExport)
 * @returns Promise<ExportProgress> - Current progress information
 * @throws Error if export not found or progress unavailable
 */
export async function getExportProgress(
  exportId: string
): Promise<ExportProgress> {
  try {
    const progress = await invoke<ExportProgress>("cmd_get_export_progress", {
      exportId,
    });
    return progress;
  } catch (error) {
    if (typeof error === "string") {
      throw new Error(error);
    }
    throw new Error(
      `Failed to get export progress: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

/**
 * Cancel an ongoing export operation
 *
 * @param exportId - ID of the export to cancel
 * @returns Promise<void>
 * @throws Error if export not found or cancellation failed
 */
export async function cancelExport(exportId: string): Promise<void> {
  try {
    await invoke("cmd_cancel_export", { exportId });
  } catch (error) {
    if (typeof error === "string") {
      throw new Error(error);
    }
    throw new Error(
      `Failed to cancel export: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}
